### PR TITLE
Refactor sessions to use access and refresh tokens

### PR DESCRIPTION
**Pull request by arkn98**
I've yet to implement a middleware to allow only unauthenticated users which is needed for routes like `ActivateUser`, `LoginUser`, etc.

### CODE REVIEW

File: internal/api/middleware/auth.go [hunk 1]
```diff
@@ -12,44 +12,35 @@ import (
...
-			usr, err := usrSvc.GetUserByID(userIDStr)
+			usr, err := userSvc.GetUserByEmpID(empID)
 			if err != nil {
 				log.Debugf("cannot get user for associated session: %v", err)
 				util.RespondWithError(w, util.ErrUnauthorized)
 				return
 			}
 
-			// store in context
+			// store user in context
```
Comment (dyaneshvarun) [inline internal/api/middleware/auth.go:43]: Agreeing that this will have more usability, but why are all fields necessary. I see this as an extra DB read operation. And at current, I think employee ID in context is more than enough for all scenarios. 

Lets assume, this is putting it in context. But what will be the flow when someone hits /user/id ?! Are we going to do the same thing again ?!

File: internal/controller/user_controller.go [hunk 2]
```diff
@@ -19,26 +20,129 @@ func UserRoutes() (string, chi.Router) {
...
+
 // LoginUser logs in an user.
 func LoginUser() http.HandlerFunc {
 	type request struct {
 		Email    string `json:"email"`
 		Password string `json:"password"`
 	}
 
+	type response struct {
+		AccessToken string `json:"access_token"`
```
Comment (dyaneshvarun) [inline internal/controller/user_controller.go:142]: Same struct used multiple times

File: internal/repo/session/session_repo.go [hunk 3]
```diff
@@ -0,0 +1,88 @@
+package session
+
```
Comment (dyaneshvarun) [inline internal/repo/session/session_repo.go:2]: Ideally should be in a different PR for this refactoring.

File: internal/service/session/session_service.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+	if err != nil {
+		log.Errorf("cannot create new refresh token: %v", err)
+		return "", "", util.ErrInternalServerError
+	}
+
+	// update session
+	if _, err := i.sessionRepo.UpdateRefreshToken(
+		s.ID,
+		hash,
+		time.Now().UTC().Add(time.Second*constants.RefreshTokenExpiry),
```
Comment (dyaneshvarun) [inline internal/service/session/session_service.go:107]: Can we have the constant name to be `RefreshTokenExpiryInSeconds`

File: internal/service/session/session_service.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+// ReadRefreshTokenCookie reads the refreshToken from the cookie.
+func (i *Impl) ReadRefreshTokenCookie(r *http.Request) (string, error) {
+	return util.GetCookie(r, constants.RefreshTokenCookieName)
+}
+
+func (i *Impl) newAccessToken(empID string) (string, error) {
+	payload := map[string]interface{}{"id": empID}
+	return crypto.NewJWT(payload, time.Second*constants.AccessTokenExpiry)
```
Comment (dyaneshvarun) [inline internal/service/session/session_service.go:179]: Same here , can we have the units specified for tokenExpiry.

File: internal/util/validation/validation.go [hunk 6]
```diff
@@ -230,6 +230,11 @@ var DayPolicy = []validation.Rule{
 	validation.Max(31),
 }
 
+// TokenPolicy is the validation policy for secure random tokens.
+var TokenPolicy = []validation.Rule{
+	validation.Required,
```
Comment (dyaneshvarun) [inline internal/util/validation/validation.go:235]: We can keep this generic if this only has Required

File: resources/create_sessions_table.sql [hunk 7]
```diff
@@ -0,0 +1,7 @@
+CREATE TABLE sessions
+(
+    id                    SERIAL PRIMARY KEY,
+    user_id               integer REFERENCES users (id),
+    refresh_token         bytea UNIQUE NOT NULL,
+    refresh_token_expires timestamp    NOT NULL
```
Comment (dyaneshvarun) [inline resources/create_sessions_table.sql:6]: This is in UTC right ?!

Comment (arkn98) [reply resources/create_sessions_table.sql:6]: It can accept any timestamp. The app should save it in UTC. We're doing that.

File: internal/service/session/session_service.go (hunk 5 above)
Comment (arkn98) [reply internal/service/session/session_service.go:179]: Sure

File: internal/service/session/session_service.go (hunk 4 above)
Comment (arkn98) [reply internal/service/session/session_service.go:107]: Okay

File: internal/api/middleware/auth.go (hunk 1 above)
Comment (arkn98) [reply internal/api/middleware/auth.go:43]: We do need other fields as well, like `email`, the surrogate key `id` (for joins), etc. As development proceeds, we may need other fields as well. 

Not everyone will be able to hit `/users/{id}`. Only the privileged will be able to. And even then they won't be requiring the details of their own account. Maybe I'll change `LoginUser` to return details of the logged-in user.

File: internal/repo/session/session_repo.go (hunk 3 above)
Comment (arkn98) [reply internal/repo/session/session_repo.go:2]: My bad

File: internal/util/validation/validation.go (hunk 6 above)
Comment (arkn98) [reply internal/util/validation/validation.go:235]: I wanted to add more rules to validate the token, like checking the length. Let me look at what can be done

File: internal/api/middleware/auth.go (hunk 1 above)
Comment (dyaneshvarun) [reply internal/api/middleware/auth.go:43]: And would be great if we can retrieve all fields from cache if available and then only hit db.

Comment (dyaneshvarun) [reply internal/api/middleware/auth.go:43]: Cos fields like email ID surrogate key are very rare to get changed for a particular Employee ID.
